Products/Zuul/catalog/reindex_solr.py

```diff
@@ -0,0 +1,320 @@
+from zenoss.modelindex.model_index import IndexUpdate, INDEX, SearchParams
+
+log = logging.getLogger('zenoss.model_catalog_reindexer')
+log.setLevel(logging.INFO)
+
+MODEL_INDEX_BATCH_SIZE = 10000
+INDEX_SIZE = 5000
+
+class WorkerReport(object):
+    def __init__(self, idx, count, err=None):
+        self.idx = idx
+        self.count = count
+        self.err = err
+
+
+class ReindexProcess(multiprocessing.Process):
+    def __init__(self, queue, idx, parent_queue, counter):
+        super(ReindexProcess, self).__init__()
+
+        self.queue = queue
+        self.idx = idx
+        self.parent_queue = parent_queue
+        self.counter = counter
+
+        self.dmd = ZenScriptBase(connect=True).dmd
+        self.index_client = zope.component.createObject('ModelIndex', get_solr_config())
+
+    def get_work(self):
+        """
+        Returns a uid for the current process to operate on.
+        Should be implemented by a subclass.
+        """
+        raise NotImplementedError
+
+    def push_children(self, node):
+        """
+        Puts the children of node in an appropriate queue.
+        Should be implemented by a subclass.
+        """
+        raise NotImplementedError
+
+    def include_node(self, node):
+        """
+        Returns whether or not to include the node in a batch to be indexed.
+        Should be implemented by a subclass.
+        """
+        raise NotImplementedError
+
+    def batch_update(self, batch, commit=False):
+        """
+        Uses a solr index client to perform a batch update of batch.
+        """
+        self.index_client.process_batched_updates(
+            [IndexUpdate(node, op=INDEX) for node in batch],
+            commit=commit)
+
+    def index(self, batch, commit=False):
+        """
+        Wraps batch_update to perform other necessary tasks.
+        Should be implemented by a subclass.
+        """
+        raise NotImplementedError
+
+    def reconcile(self):
+        """
+        Reconciles queues or other properties after operating on a node.
+        Should be implemented by a subclass.
+        """
+        raise NotImplementedError
+
+    def do_work(self):
+        """
+        Operates on uids/nodes from get_work until exhausted.
+        """
+        count = 0
+        batch = []
+
+        try:
+            for current_uid in self.get_work():
+                try:
+                    current_node = self.dmd.unrestrictedTraverse(current_uid)
+                except NotFound:
+                    log.error("'{0}' was not found in ZODB.  A hard reindex or investigation may be needed.".format(current_uid))
+                    continue
+                self.push_children(current_node)
+                if self.include_node(current_node):
+                    batch.append(current_node)
+                    count += 1
+                    # the increment on counter is non-atomic, but we don't care
+                    self.counter.value += 1
```

Comment:
Shouldn't we also pass a lock to the process and use it? Also, it would be probably better to update the lock every, let's say, 1k elements.
I prefer the approach of having the worker send a msg to the main process (we have the queue already set up) instead of using a shared variable.